### ISSUE TITLE
Low memory limit causes OCI runtime create failed when running container

### ISSUE THREAD

**Issue by rabbitfang** [moby/moby#38921]
**Description**
Attempting to start a container with a low memory limit fails

**Steps to reproduce the issue:**
1. `docker run --memory 5m --memory-swap=5m hello-world`

**Describe the results you received:**
```
docker: Error response from daemon: OCI runtime create failed: container_linux.go:344: starting container process caused "process_linux.go:424: container init caused \"process_linux.go:390: setting cgroup config for procHooks process caused \\\"failed to write 5242880 to memory.memsw.limit_in_bytes: write /sys/fs/cgroup/memory/docker/0e4e8c8e8b7343ffcbdb90d6c9c96a60aa49fb54a288dfe75b649b2a0c370038/memory.memsw.limit_in_bytes: device or resource busy\\\"\"": unknown.
ERRO[0000] error waiting for container: context canceled
```

or sometimes

```
docker: Error response from daemon: cannot start a stopped process: unknown.
ERRO[0001] error waiting for container: context canceled
```

**Describe the results you expected:**
The hello world image runs successfully.

**Additional information you deem important (e.g. issue happens only occasionally):**
- This error does not happen every time. It happens nearly every time with ` --memory-swappiness=0` also being passed.
- Increasing the value of `--memory` and `--memory-limit` decreases the rate at which they occur. It nearly always occurs at `5m`, and never occurs at `20m`.
- [The docs](https://docs.docker.com/engine/reference/run/#runtime-constraints-on-resources) specify a lower bound for the `--memory` arg at 4M
- If you separate out `docker run` into `docker create` and `docker start`, the error happens during `docker start`.

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           18.09.2
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        6247962
 Built:             Sun Feb 10 04:12:39 2019
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.2
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.6
  Git commit:       6247962
  Built:            Sun Feb 10 04:13:06 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
Containers: 4
 Running: 0
 Paused: 0
 Stopped: 4
Images: 15
Server Version: 18.09.2
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9754871865f7fe2f4e74d43e2fc7ccd237edcbce
runc version: 09c8266bf2fcf9519a651b04ae54c967b9ab86ec
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.125-linuxkit
Operating System: Docker for Mac
OSType: linux
Architecture: x86_64
CPUs: 3
Total Memory: 1.952GiB
Name: linuxkit-025000000001
ID: JYIU:2M5K:5AN5:UBXW:CHND:HRCS:OHJV:MWXQ:BJYS:RA6A:ZV5C:ZDHW
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 26
 Goroutines: 53
 System Time: 2019-03-21T17:22:10.771174637Z
 EventsListeners: 2
HTTP Proxy: gateway.docker.internal:3128
HTTPS Proxy: gateway.docker.internal:3129
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
The data above is from a Docker for Mac environment, but it also occurs on Ubuntu 18.04 with Docker CE 18.06.3 (Kubernetes node).

**Comment by rabbitfang** [comment]
I tested on a machine running Docker CE 17.06.2 and Ubuntu 16.04 and was not able to reproduce the issue.

**Comment by cpuguy83** [comment]
@rabbitfang This is very likely an issue with the way CVE-2019-5736 has been fixed... basically it ends up using an additional ~10MB of memory... this happens because the fix makes an in-memory copy of itself.

This should be fixed in newer versions of runc... not sure if this is included in docker 18.09.3

**Comment by cpuguy83** [comment]
I'm going to close this as it should be a non-issue with the fix that just needs to trickle down to a Docker CE release.

Note that docker sets a hard lower limit of 4MB (which may need to be re-addressed) because the memory limit needs to account for the amount of memory that runc itself takes.

**Comment by rabbitfang** [comment]
Upstream runc issue: https://github.com/opencontainers/runc/issues/1980

**Comment by thaJeztah** [comment]
Updated version of runc with the reduced memory use is included in containerd 1.2.5 (https://github.com/containerd/containerd/releases/tag/v1.2.5), which has not shipped yet as a package on download.docker.com, but will be released soon (it's already available in the "test" channel; https://download.docker.com/linux/ubuntu/dists/xenial/pool/test/amd64/)

**Comment by rabbitfang** [comment]
@thaJeztah, thanks. I have already worked around it by building a custom version of the docker-runc binary (using [your PR](https://github.com/docker/runc/pull/8)) when installing Docker. It's not the prettiest thing to do, but it works well enough.
